unittests/runtime/Metadata.cpp

```diff
@@ -21,127 +21,75 @@
-  if (count == 0) {
-    errno = EINVAL;
-    return -1;
-  }
-  if (pthread_mutex_init(&barrier->mutex, nullptr) != 0) {
-    return -1;
-  }
-  if (pthread_cond_init(&barrier->cond, nullptr) != 0) {
-    pthread_mutex_destroy(&barrier->mutex);
-    return -1;
-  }
-  barrier->count = count;
-  barrier->numThreadsWaiting = 0;
-  return 0;
-}
-
-static int pthread_barrier_destroy(pthread_barrier_t *barrier) {
-  // want to destroy both even if destroying one fails.
-  int ret = 0;
-  if (pthread_cond_destroy(&barrier->cond) != 0) {
-    ret = -1;
-  }
-  if (pthread_mutex_destroy(&barrier->mutex) != 0) {
-    ret = -1;
-  }
-  return ret;
-}
-
-static int pthread_barrier_wait(pthread_barrier_t *barrier) {
-  if (pthread_mutex_lock(&barrier->mutex) != 0) {
-    return -1;
-  }
-  ++barrier->numThreadsWaiting;
-  if (barrier->numThreadsWaiting < barrier->count) {
-    // Put the thread to sleep.
-    if (pthread_cond_wait(&barrier->cond, &barrier->mutex) != 0) {
-      return -1;
-    }
-    if (pthread_mutex_unlock(&barrier->mutex) != 0) {
-      return -1;
-    }
-    return 0;
-  } else {
-    // Reset thread count.
-    barrier->numThreadsWaiting = 0;
-
-    // Wake up all threads.
-    if (pthread_cond_broadcast(&barrier->cond) != 0) {
-      return -1;
-    }
-    if (pthread_mutex_unlock(&barrier->mutex) != 0) {
-      return -1;
-    }
-    return PTHREAD_BARRIER_SERIAL_THREAD;
-  }
-}
-#endif
+#include <thread>
+#include <condition_variable>
 
 using namespace swift;
 
 // Race testing.
 
 template <typename T>
-struct RaceArgs {
+struct RaceThreadContext {
   std::function<T()> code;
-  pthread_barrier_t *go;
+  T result;
+
+  unsigned numThreads;
+  unsigned &numThreadsReady;
+  std::mutex &sharedMutex;
+  std::condition_variable &start_condition;
 };
 
-void *RaceThunk(void *vargs) {
-  RaceArgs<void*> *args = static_cast<RaceArgs<void*> *>(vargs);
-  // Signal ready. Wait for go.
-  pthread_barrier_wait(args->go);
-  return args->code();
+template <typename T>
+void RaceThunk(RaceThreadContext<T> ctx) {
+  // update shared state
+  std::unique_lock<std::mutex> lk(ctx.sharedMutex);
+  ++ctx.numThreadsReady;
+  bool isLastThread = ctx.numThreadsReady == ctx.numThreads;
+
+  // wait until the rest of the thunks are ready
+  ctx.start_condition.wait(lk, [&ctx]{ // waiting releases the lock
+    return ctx.numThreadsReady == ctx.numThreads;
+  });
+  lk.unlock();
+
+  // The last thread will signal the condition_variable to kick off the rest
+  // of the waiting threads to start.
+  if (isLastThread) ctx.start_condition.notify_all();
+
+  ctx.result = ctx.code();
 }
 
 /// RaceTest(code) runs code in many threads simultaneously, 
 /// and returns a vector of all returned results.
 template <typename T, int NumThreads = 64>
-std::vector<T> 
-RaceTest(std::function<T()> code)
-{
+std::vector<T> RaceTest(std::function<T()> code) {
   const unsigned threadCount = NumThreads;
```

Comment:
Im not sure I understand the point of the `threadCount` variable.  Why not just use `NumThreads` everywhere?

Reply:
That was in the original implementation, I think it was because pthreads required an `unsigned`.
I'll remove it